xds/internal/balancer/xdsrouting/routing_config.go

```diff
@@ -0,0 +1,205 @@
+	exactMatch, regexMatch, prefixMatch, suffixMatch string
+	rangeMatch                                       *int64Range
+	presentMatch                                     bool
+}
+
+type route struct {
+	// Path, Prefix and Regex can have at most one set. This is guaranteed by
+	// config parsing.
+	path, prefix, regex string
+
+	headers  []headerMatcher
+	fraction *uint32
+
+	// Action is the name from the action list.
+	action string
+}
+
+// lbConfig is the balancer config for xds routing policy.
+type lbConfig struct {
+	serviceconfig.LoadBalancingConfig
+	routes  []route
+	actions map[string]action
+}
+
+// The following structs with `JSON` in name are temporary structs to unmarshal
+// json into. The fields will be read into lbConfig, to be used by the balancer.
+
+// routeJSON is temporary struct for json unmarshal.
+type routeJSON struct {
+	// Path, Prefix and Regex can have at most one non-nil.
+	Path, Prefix, Regex *string
+	// Zero or more header matchers.
+	Headers       []*xdsclient.HeaderMatcher
+	MatchFraction *wrapperspb.UInt32Value
+	// Action is the name from the action list.
+	Action string
+}
+
+// lbConfigJSON is temporary struct for json unmarshal.
+type lbConfigJSON struct {
+	Route  []routeJSON
+	Action map[string]action
+}
+
+func (jc lbConfigJSON) toLBConfig() *lbConfig {
+	var ret lbConfig
+	for _, r := range jc.Route {
+		var tempR route
+		switch {
+		case r.Path != nil:
+			tempR.path = *r.Path
+		case r.Prefix != nil:
+			tempR.prefix = *r.Prefix
+		case r.Regex != nil:
+			tempR.regex = *r.Regex
+		}
+		for _, h := range r.Headers {
+			if h.RangeMatch != nil {
+				fmt.Println("range not nil", *h.RangeMatch)
+			}
+			var tempHeader headerMatcher
+			switch {
+			case h.ExactMatch != nil:
+				tempHeader.exactMatch = *h.ExactMatch
+			case h.RegexMatch != nil:
+				tempHeader.regexMatch = *h.RegexMatch
+			case h.PrefixMatch != nil:
+				tempHeader.prefixMatch = *h.PrefixMatch
+			case h.SuffixMatch != nil:
+				tempHeader.suffixMatch = *h.SuffixMatch
+			case h.RangeMatch != nil:
+				tempHeader.rangeMatch = &int64Range{
+					start: h.RangeMatch.Start,
+					end:   h.RangeMatch.End,
+				}
+			case h.PresentMatch != nil:
+				tempHeader.presentMatch = *h.PresentMatch
+			}
+			tempHeader.name = h.Name
+			if h.InvertMatch != nil {
+				tempHeader.invertMatch = *h.InvertMatch
+			}
+			tempR.headers = append(tempR.headers, tempHeader)
+		}
+		if r.MatchFraction != nil {
+			tempR.fraction = &r.MatchFraction.Value
+		}
+		tempR.action = r.Action
+		ret.routes = append(ret.routes, tempR)
+	}
+	ret.actions = jc.Action
+	return &ret
+}
+
+func parseConfig(c json.RawMessage) (*lbConfig, error) {
+	var tempConfig lbConfigJSON
+	if err := json.Unmarshal(c, &tempConfig); err != nil {
+		return nil, err
+	}
+
+	// For each route:
+	// - at most one of path/prefix/regex.
+	// - action is in action list.
+
+	allRouteActions := make(map[string]bool)
+	for _, r := range tempConfig.Route {
```

Comment:
All of this is a lot of manual parsing and validation. Was using something like https://godoc.org/github.com/golang/protobuf/jsonpb considered?

If we add the LB config definition in the design doc to grpc-proto repo, we can import here and directly convert from proto to JSON, can't we?

Reply:
The balancer doesn't deal with proto messages. It doesn't import the proto at all. jsonpb only works with proto messages I think.

And, none of these will exist after we implement route actions. xds resolver will consume the structs from xds client directly. Instead of marshalling into json, and giving to balancers.